### ISSUE TITLE
HTTPS by default

### ISSUE THREAD

**Issue by konklone** [18F/api.data.gov#34]
HTTPS is already configured at [https://api.data.gov](https://api.data.gov/). So, api.data.gov should make it the recommended and default protocol.

I'm sure the team here already understands the various benefits of HTTPS, and I don't mean to preach to the choir, but here are a few:

* Broader client-side compatibility. Any APIs under api.data.gov wishing to support CORS or JSONP for in-browser operation will be completely blocked on HTTPS websites, as active mixed content.
* Enhanced privacy for apps and users using APIs at api.data.gov -- HTTP headers and query string parameters (among other things) will be encrypted. For APIs accepting particularly sensitive data (like lat/long or zip), [this is critical](http://www.theguardian.com/world/2014/jan/27/nsa-gchq-smartphone-app-angry-birds-personal-data).
* The primary benefit of HTTPS: a stronger guarantee that you're speaking with the real api.data.gov.

Much of the performance hit involved in SSL can be mitigated - @igrigorik's excellent [istlsfastyet.com](https://istlsfastyet.com/) contains many references for doing this.

There are a few ways this issue could be tackled:

* Update the documentation to always use `https://` in the examples.
* Update all links to api.data.gov to use `https://`.
* Begin redirecting all `http://` requests to `https://` for the main `api.data.gov` website.
* Begin redirecting all `http://` requests to `https://` for any contained API.
* Set a [HSTS header](https://en.wikipedia.org/wiki/HTTP_Strict_Transport_Security) that tells clients to always use `https://` going forward.

**Comment by konklone** [comment]
This thread was the inspiration for some of the content that @nacin wrote up for this project: https://https.cio.gov/apis/

As api.data.gov migrates, feeding some of the experience back into that page, managed for now at [whitehouse/https](https://github.com/WhiteHouse/https/), would be awesome for other agencies looking to do something similar.

**Comment by GUI** [comment]
I saw that API page when the HTTPS site got released and thought it sounded vaguely familiar to our conversation here. :) But awesome work to everyone involved in the HTTPS standard--it looks really great.

But as an extremely belated followup, we are *finally* planning to at least tackle part of this for api.data.gov. In the next couple weeks we're planning to rollout functionality that will force all new APIs we host to HTTPS (and you can thank the HTTPS standard site going live for jogging my memory). Even though I've been trying to informally encourage all new users to use HTTPS, it easily gets missed as soon as people start exchanging links with HTTP in them. This should at least help us stem the tide of non-HTTPS APIs and ensure all new agencies moving forward on api.data.gov will be HTTPS-only (I just wish we had tackled this earlier). I may also try to implement something to try to ensure that new API keys that are signed up for are only used via HTTPS, even for existing APIs (but I'm less sure about how feasible that will be to implement).

Then the second part of this task becomes migrating existing APIs to HTTPS-only. As we've discussed, that's a little more complicated and will require a little more coordination/communication, but it's certainly doable. But we at least wanted to start with the simpler task of ensuring new clients are HTTPS-only, and then we can begin to work with existing agencies on potential migration.

And yeah, I'd definitely be happy to share our experiences and things we learn.

Thanks again for everything you're doing to push HTTPS forward, and sorry for the super-long delay in following up!

**Comment by konklone** [comment]
> In the next couple weeks we're planning to rollout functionality that will force all new APIs we host to HTTPS (and you can thank the HTTPS standard site going live for jogging my memory).

*Awesome.*

> I may also try to implement something to try to ensure that new API keys that are signed up for are only used via HTTPS, even for existing APIs (but I'm less sure about how feasible that will be to implement).

That's fascinating. If that is feasible, that's exactly the kind of case study I'd like to document.

> Then the second part of this task becomes migrating existing APIs to HTTPS-only. As we've discussed, that's a little more complicated and will require a little more coordination/communication, but it's certainly doable. But we at least wanted to start with the simpler task of ensuring new clients are HTTPS-only, and then we can begin to work with existing agencies on potential migration.

Absolutely. I don't underestimate the challenge. The kinds of steps you think of will also be useful.

> Thanks again for everything you're doing to push HTTPS forward, and sorry for the super-long delay in following up!

Thank you for being so supportive of the transition!

**Comment by GUI** [comment]
This has been rolled out to production. Any new API backends we setup will default to HTTPS-only access. All existing APIs continue to be accessible over HTTP or HTTPS, but I've introduced a couple new "transitionary" options that can be applied to individual API backends to hopefully aid in eventually switching these over to HTTPS.

The transition options allow for any existing API keys to continue accessing APIs over HTTP (to prevent breaking usage for current users). However, any new API keys created after the transition mode is put into place on an API will be forced to use HTTPS. As a separate task, I think we can talk with existing agencies about at least putting their APIs into this transition mode. It should be a very safe option, since it doesn't affect existing users, the only reason I didn't default everyone to this already is that I think we should talk to agencies about this first and help ensure that their docs are using HTTPS links. This will just ensure new users have a smooth experience and don't encounter docs referencing insecure HTTP links, which would results in possible errors for the new users.

When picking your API's HTTPS requirements, you can also pick whether to redirect http users to https, or whether to return a straight-up error to http users. While redirects would technically allow for a smoother transition in some cases, there's a few reasons I think redirects aren't ideal for API usage. So for new APIs, I'm defaulting to an error message for HTTP access that instructs the user to use an HTTPS link instead.

Here's the help we provide in the admin that explains the options and some of the reasoning in a bit more detail:

> Choose whether HTTPS is required to access this API. HTTPS is encouraged to protect the API keys.
> 
> - **Required & return message:** HTTPS is required to access the API. HTTP requests will return an error message instructing the user to use an HTTPS URL instead. This is the recommended and default strategy.
> - **Required & return redirect:** HTTPS is required to access the API. HTTP requests will return a redirect to the HTTPS URL.
> - **Transitionary & return message:** New API keys that signup after choosing this setting will be forced to use HTTPS. Existing API keys may continue to use either HTTP or HTTPS. New API keys using HTTP will return an error message instructing the user to use an HTTPS URL instead.
> - **Transitionary & return redirect:** New API keys that signup after choosing this setting will be forced to use HTTPS. Existing API keys may continue to use either HTTP or HTTPS. New API keys using HTTP will return a redirect to the HTTPS URL.
> - **Optional**: HTTPS is optional and either HTTP or HTTPS may be used.
> 
> *Notes on redirects:*
> 
> - Not all API clients will automatically follow redirects, so be careful if using a redirect strategy for existing APIs (since existing calls may break).
> - If API clients rely on the redirect for HTTPS access, this strategy does not secure the API keys, since the client may still be making an insecure initial HTTP request with their API key.
> - For GET requests a `301 Moved Permanently` redirect will be returned. For all other HTTP methods a `307 Temporary Redirect` redirect will be returned (to instruct the client to retry using the same HTTP method).

As I mentioned, I think we still have some work to do to transition existing agencies to HTTPS, but I think we can tackle that as a separate issue. Let me know if anyone has any other thoughts on any of this or on some of the implementation details.

Implemented in https://github.com/NREL/api-umbrella-gatekeeper/pull/10, https://github.com/NREL/api-umbrella-web/pull/9, and https://github.com/NREL/api-umbrella-router/pull/5

**Comment by konklone** [comment]
> This has been rolled out to production. Any new API backends we setup will default to HTTPS-only access. 

:clap: :clap: :clap: :clap: :clap: :clap: :lock :lock: :lock :lock: :lock :lock: :confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball: :confetti_ball: 

That's so great, @GUI! And this is thorough work.

> All existing APIs continue to be accessible over HTTP or HTTPS, but I've introduced a couple new "transitionary" options that can be applied to individual API backends to hopefully aid in eventually switching these over to HTTPS. The transition options allow for any existing API keys to continue accessing APIs over HTTP (to prevent breaking usage for current users). However, any new API keys created after the transition mode is put into place on an API will be forced to use HTTPS. 

That's so smart. It effectively means that the "surface area" of clients that APIs need to push to migrate can be immediately capped, even as the actual transition process might drag out over the long term.

> As a separate task, I think we can talk with existing agencies about at least putting their APIs into this transition mode. It should be a very safe option, since it doesn't affect existing users, the only reason I didn't default everyone to this already is that I think we should talk to agencies about this first and help ensure that their docs are using HTTPS links. This will just ensure new users have a smooth experience and don't encounter docs referencing insecure HTTP links, which would results in possible errors for the new users.

Makes total sense.

> When picking your API's HTTPS requirements, you can also pick whether to redirect http users to https, or whether to return a straight-up error to http users. While redirects would technically allow for a smoother transition in some cases, there's a few reasons I think redirects aren't ideal for API usage. So for new APIs, I'm defaulting to an error message for HTTP access that instructs the user to use an HTTPS link instead.

Also makes total sense.

> As I mentioned, I think we still have some work to do to transition existing agencies to HTTPS, but I think we can tackle that as a separate issue. Let me know if anyone has any other thoughts on any of this or on some of the implementation details.

:+1: Feel free to tag me on any other issues you open around this. I'll be trying to translate api.data.gov's transition experience into recommendations for other agencies.

**Comment by nacin** [comment]
@GUI, this is fantastic. I love the idea about using API keys to control HTTPS-only access, this is so well done.

And now for a rant about redirects. :-)

I think redirects should be avoided if at all possible. https://https.cio.gov/apis/ goes into very specific detail as to why it's best to simply not use them, so I won't rehash what's already there. I would very much hesitate making it an option — I don't think it's a decision that can easily be made by choosing from a menu of options.

I do think there are limited situations where a redirect can be OK. Broadly, that'd be CORS, JSONP, etc. Narrowly, I do agree you can redirect read-only APIs if you're willing to accept the risk, as you basically lay out here. But I wouldn't want to emphasize that — it's more likely than not too risky. Any 30x is going to break some clients that aren't following redirects; any HTTPS URL might break some clients that aren't configured properly for HTTPS; a 307 is going to break clients that don't understand it; etc. I know this is stated in the notes (Is it worth linking these notes to https://https.cio.gov/apis/?), but I'd suggest even greater de-emphasis on redirects as a sub-par solution compared to the other options.

The nuance here is why https://https.cio.gov/apis/ simply says "don't use them" – it's easier to have blanket guidance, the same way the HTTPS-only proposal does not make exceptions for so-called non-sensitive content.

I would also strike the "Transitionary & return redirect" option as, as you noted, it's inherently insecure. If their HTTPS-only API key never works over HTTP (as in, returns an error instead of a redirect), then they won't accidentally do something that leaks the API key over HTTP (perhaps only development when trying it out).

**Comment by nacin** [comment]
Also — I would love to expand https://https.cio.gov/apis/ to talk more about API keys and specifically the tactics outlined here.

**Comment by konklone** [comment]
> Also — I would love to expand https://https.cio.gov/apis/ to talk more about API keys and specifically the tactics outlined here.

Me too! While we should probably avoid changing https.cio.gov much during the comment period, I know I'm making a private branch or two with some other work I think should be added to the site after the comment period closes. We can share with each other over ngrok or whatever in the meantime, too.

**Comment by GUI** [comment]
@nacin: Thanks for the feedback! I agree with everything you've said about redirects. I added the redirect support with the general thought that it might theoretically help as a transition path towards HTTPS for specific APIs, but the more I think about it, and after reading your comments, I think you're right--redirects don't really help matters much, and probably just confuse things.

We could certainly remove the redirect options altogether at this point. Can anyone think of any reasons to keep either redirect option (required or transitionary) at all?

If we do want to remove those options, it would also probably simplify the options for admins and prevent them from taking a path that we don't really want to encourage. If we remove things, the new simplified options would be:

> - **Required:** HTTPS is required to access the API. HTTP requests will return an error message instructing the user to use an HTTPS URL instead. This is the recommended and default strategy.
> - **Transitionary:** New API keys that signup after choosing this setting will be forced to use HTTPS. Existing API keys may continue to use either HTTP or HTTPS. New API keys using HTTP will return an error message instructing the user to use an HTTPS URL instead.
> - **Optional**: HTTPS is optional and either HTTP or HTTPS may be used.

**Comment by konklone** [comment]
> Can anyone think of any reasons to keep either redirect option (required or transitionary) at all?

The only reason I can think of would be an API with heavy existing CORS/JSONP usage. However, you can address that more straightforwardly by adding an HSTS header to api.data.gov. 

And I think you can add that HSTS header pretty much any time, as browsers are the only API clients that will obey HSTS headers, and by obeying them they should Just Work and only request HTTPS.

**Comment by konklone** [comment]
> However, you can address that more straightforwardly by adding an HSTS header to api.data.gov.

...and to APIs using custom domain names, as long as those custom domain names have HTTPS enabled.
